File: easy-projects/easy-common/src/main/java/xyz/srclab/common/reflect/FieldKit.java
package xyz.srclab.common.reflect;

import xyz.srclab.annotation.Immutable;
import xyz.srclab.annotation.Nullable;
import xyz.srclab.common.collection.ListKit;

import java.lang.reflect.Field;
import java.util.List;

public class FieldKit {

    @Nullable
    public static Field getField(Class<?> cls, String name) {
        try {
            return cls.getField(name);
        } catch (NoSuchFieldException e) {
            return null;
        }
    }

    @Nullable
    public static Field getDeclaredField(Class<?> cls, String name) {
        try {
            return cls.getDeclaredField(name);
        } catch (NoSuchFieldException e) {
            return null;
        }
    }

    @Immutable
    public static List<Field> getFields(Class<?> cls) {
        return ListKit.immutable(cls.getFields());
    }

    @Immutable
    public static List<Field> getDeclaredFields(Class<?> cls) {
        return ListKit.immutable(cls.getDeclaredFields());
    }

    @Nullable
    public static Object getValue(Field field, @Nullable Object object) {
        return getValue(field, object, false);
    }

    public static void setValue(Field field, @Nullable Object object, @Nullable Object value) {
        setValue(field, object, value, false);
    }

    @Nullable
    public static Object getValue(Field field, @Nullable Object object, boolean force) {
        try {
            if (force) {
                field.setAccessible(true);
            }
            return field.get(object);
        } catch (IllegalAccessException e) {
            throw new IllegalStateException(e);
        }
    }

    public static void setValue(Field field, @Nullable Object object, @Nullable Object value, boolean force) {
        try {
            if (force) {
                field.setAccessible(true);
            }
            field.set(object, value);
        } catch (IllegalAccessException e) {
            throw new IllegalStateException(e);
        }
    }
}

File: srclab-common/src/main/kotlin/xyz/srclab/common/reflect/FieldKit.kt
@file:JvmName("FieldKit")
@file:JvmMultifileClass

package xyz.srclab.common.reflect

import xyz.srclab.common.base.asAny
import java.lang.reflect.Field

@JvmOverloads
fun Class<*>.findField(name: String, declared: Boolean = false, deep: Boolean = false): Field? {
    var field = try {
        this.getField(name)
    } catch (e: NoSuchFieldException) {
        null
    }
    if (field === null && !declared) {
        return null
    }
    field = this.findDeclaredField(name)
    if (field === null && !deep) {
        return null
    }
    var superClass = this.superclass
    while (superClass !== null) {
        field = this.findDeclaredField(name)
        if (field !== null) {
            return field
        }
        superClass = superclass.superclass
    }
    return null
}

fun Class<*>.findDeclaredField(name: String): Field? {
    return try {
        this.getDeclaredField(name)
    } catch (e: NoSuchFieldException) {
        null
    }
}

fun Class<*>.findFields(): List<Field> {
    return this.fields.asList()
}

fun Class<*>.findDeclaredFields(): List<Field> {
    return this.declaredFields.asList()
}

@JvmOverloads
fun <T> Field.getValue(owner: Any? = null, force: Boolean = false): T {
    return try {
        if (force) {
            this.isAccessible = true
        }
        this.get(owner).asAny()
    } catch (e: IllegalAccessException) {
        throw IllegalStateException(e)
    }
}

@JvmOverloads
fun Field.setValue(value: Any?, owner: Any? = null, force: Boolean = false) {
    try {
        if (force) {
            this.isAccessible = true
        }
        this.set(owner, value)
    } catch (e: IllegalAccessException) {
        throw IllegalStateException(e)
    }
}

@JvmOverloads
fun <T> Class<*>.getFieldValue(name: String, owner: Any? = null, force: Boolean = false, deep: Boolean = false): T {
    val field = this.findField(name, force, deep)
    if (field === null) {
        throw IllegalArgumentException("Field not found: ${this}.${name}")
    }
    return field.getValue(owner, force)
}

@JvmOverloads
fun Class<*>.setFieldValue(
    name: String,
    value: Any?,
    owner: Any? = null,
    force: Boolean = false,
    deep: Boolean = false
) {
    val field = this.findField(name, force, deep)
    if (field === null) {
        throw IllegalArgumentException("Field not found: ${this}.${name}")
    }
    return field.setValue(value, owner, force)
}